### PR TITLE
Ensure survey prompt is not displayed multiple times

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #4002

-   [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [ ] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/client/datascience/dataScienceSurveyBanner.ts [hunk 1]
```diff
@@ -77,21 +80,31 @@ export class DataScienceSurveyBannerLogger implements IInteractiveWindowListener
...
+        if (this.applicationEnvironment.uiKind !== UIKind.Desktop || this.applicationEnvironment.channel !== 'stable') {
+            return false;
+        }
+        if (!this.showBannerState.value.enabled) {
+            return false;
+        }
+        if (!this.showBannerState.value.expiryTimeInMs) {
+            return this.showBannerState.value.enabled;
+        }
+        return this.showBannerState.value.expiryTimeInMs! < Date.now();
```
Comment (DonJayamanne) [inline src/client/datascience/dataScienceSurveyBanner.ts:99]: Explicit check for expiry

File: src/client/datascience/dataScienceSurveyBanner.ts [hunk 2]
```diff
@@ -77,21 +79,25 @@ export class DataScienceSurveyBannerLogger implements IInteractiveWindowListener
...
-            env.uiKind !== UIKind?.Web &&
-            this.applicationEnvironment.channel === 'stable'
-        );
+        // if (this.applicationEnvironment.uiKind !== UIKind.Desktop || this.applicationEnvironment.channel !== 'stable') {
+        //     return false;
+        // }
+        if (!this.showBannerState.value.expiry) {
+            return true;
+        }
+        return this.showBannerState.value.expiry! < Date.now();
```
Comment (DonJayamanne) [inline src/client/datascience/dataScienceSurveyBanner.ts:92]: Explicit check for expiry.
This is the structure of the data stored in persistent storage if it contains any expiry information (hence ensuring we keep using the old data).

File: src/client/datascience/dataScienceSurveyBanner.ts [hunk 3]
```diff
@@ -149,12 +159,11 @@ export class DataScienceSurveyBanner implements IJupyterExtensionBanner {
...
     }
-
     private async disable(monthsTillNextPrompt: number) {
-        const expiration = monthsTillNextPrompt * 31 * 24 * 60 * 60 * 1000;
-        await this.persistentState
-            .createGlobalPersistentState<boolean>(DSSurveyStateKeys.ShowBanner, false, expiration)
-            .updateValue(false);
+        await this.showBannerState.updateValue({
+            expiry: monthsTillNextPrompt * 31 * MillisecondsInADay,
+            data: true
```
Comment (DonJayamanne) [inline src/client/datascience/dataScienceSurveyBanner.ts:165]: Take over expiry as it can change over time.

Comment (IanMatthewHuff) [inline src/client/datascience/dataScienceSurveyBanner.ts:165]: I think we want false and not true here in disable.

Comment (DonJayamanne) [reply src/client/datascience/dataScienceSurveyBanner.ts:165]: Doesn't get used, I'm looking at the `expiry` value instead.

Comment (DonJayamanne) [reply src/client/datascience/dataScienceSurveyBanner.ts:165]: Basically if there is no `expiry` then its enabled, else its disabled based on expiry value.

Comment (IanMatthewHuff) [reply src/client/datascience/dataScienceSurveyBanner.ts:165]: Ahh got it now, I see that in the enabled function implementation. Yeah, that works. I might toss a comment here at this point? Without looking at the enabled implementation it's a bit non-intuitive.

Comment (IanMatthewHuff) [reply src/client/datascience/dataScienceSurveyBanner.ts:165]: @DonJayamanne Wait hang on one more second. Don't you need to add a Date.now() onto the value here? disable(1) will be the same value today and tomorrow. Persistent state expiry did that under the surface.

Comment (DonJayamanne) [reply src/client/datascience/dataScienceSurveyBanner.ts:165]: Thanks

Comment (DonJayamanne) [reply src/client/datascience/dataScienceSurveyBanner.ts:178]: Phew, big bug, thanks. Tests pass as its a relative time.

File: src/client/datascience/dataScienceSurveyBanner.ts [hunk 4]
```diff
@@ -77,21 +79,38 @@ export class DataScienceSurveyBannerLogger implements IInteractiveWindowListener
...
+    /**
+     * This value is not used.
+     * We are only interested in the value for `expiry`.
+     * This structure is based on the old data for older customers when we used PersistentState class.
+     */
+    data: boolean;
+    /**
+     * If this is value `undefined`, then prompt can be displayed.
+     * If this value is `a number`, then a prompt was displayed at one point in time &
+     * we need to wait for Date.now() to be greater than that number to display it again.
```
Comment (DonJayamanne) [inline src/client/datascience/dataScienceSurveyBanner.ts:92]: @IanMatthewHuff added comments here.

File: src/client/datascience/dataScienceSurveyBanner.ts (hunk 3 above)
Comment (IanMatthewHuff) [reply src/client/datascience/dataScienceSurveyBanner.ts:178]: Yeah, that's the tricky bit with tests like that. You don't want tests to wait a month so you gotta fake it up somehow :).

**Review by IanMatthewHuff — APPROVED** [review]

**Comment by joyceerhl** [comment]
I think we'd need this fix in the insidersNativeNotebookSurvey.ts too.

**Comment by DonJayamanne** [comment]
Dmn it, just merged the PR.
Will create another PR for that.

**Pull request merged**